Return Result to fail gracefully instead of aborting

lucylq opened the pull request:
Summary:
- Make TensorInfo static factory pattern to allow returning Result.
- Return Result type to fail gracefully instead of directly aborting in the runtime.

Differential Revision: D78187751

JacobSzwejbka commented on runtime/executor/method_meta.cpp:
```diff
@@ -52,7 +52,7 @@ Result<Tag> get_tag(
 }
 
-size_t calculate_nbytes(
+Result<size_t> calculate_nbytes(
```
You probably have to leave the old one and mark it deprecated. It would be good to do it for every thing in TensorInfo that doesnt return result

lucylq replied:
TensorInfo constructor is private (only MethodMeta can create it). Should be OK?

lucylq replied:
I think this is the only one in TensorInfo. The rest of the functions are getters that don't error-check.

lucylq committed:
Return Result to fail gracefully instead of aborting (#12409)

Summary:

- Make TensorInfo static factory pattern to allow returning Result.
- Return Result type to fail gracefully instead of directly aborting in the runtime.

Reviewed By: JacobSzwejbka

Differential Revision: D78187751
